### PR TITLE
Emit duplicate target dependency warnings

### PR DESCRIPTION
**Pull request by michaelmcguire**
Fixes: https://bugs.swift.org/browse/SR-9402

### CODE REVIEW

File: Sources/PackageGraph/PackageGraphLoader.swift [hunk 1]
```diff
@@ -530,6 +544,19 @@ private final class ResolvedTargetBuilder: ResolvedBuilder<ResolvedTarget> {
             deps.append(.product(product))
         }
 
+        // Find duplicate target dependencies and emit a warning
+        deps.reduce(into: [ResolvedTarget.Dependency: Int]()) { (
```
[the remark this answers is not in the document]
Comment (michaelmcguire) [reply Sources/PackageGraph/PackageGraphLoader.swift]: I'm new to Swift and SPM development. For these changes is an additional commit ok or should I be rewriting history so this PR only has one commit?

[the remark this answers is not in the document]
Comment (michaelmcguire) [reply Sources/PackageGraph/PackageGraphLoader.swift]: I remember now why I ended up doing this in `PackageGraphLoader` and not in `PackageBuilder`. By the end of the `PackageBuilder.construct` method, the dependencies are not fully resolved. I could look at `Target.dependencies` and `Target.productDependencies` to look for duplicates, but technically things could fall through cracks (i.e. you declared one product dependency `byName` and another one fully with both `name` and `product`). Here's an example of what I mean:

```
// swift-tools-version:4.2

import PackageDescription

let package = Package(
    name: "Example",
    products: [
        .library(
            name: "Example",
            targets: ["Example"]),
    ],
    dependencies: [
        .package(url: "https://github.com/behrang/YamlSwift.git", from: "3.4.3")
    ],
    targets: [
        .target(
            name: "Example",
            dependencies: ["Yaml", .product(name: "Yaml", package: "Yaml")]),
    ]
)
```

I don't _think_ that duplicate target dependency of `Yaml` can be seen in `PackageBuilder`.

[the remark this answers is not in the document]
Comment (aciidb0mb3r) [reply Sources/PackageGraph/PackageGraphLoader.swift]: The names need to be unique anyway so we should emit an error in PackageBuilder itself.

[the remark this answers is not in the document]
Comment (aciidb0mb3r) [reply Sources/PackageGraph/PackageGraphLoader.swift]: >I'm new to Swift and SPM development. For these changes is an additional commit ok or should I be rewriting history so this PR only has one commit?
 
I prefer rewriting the history but it's not a big deal as maintainers can use the "squash and merge" option when merging.

**Comment by michaelmcguire** [comment]
@aciidb0mb3r Moved the checks to `PackageBuilder(Tests)`, ready for review.

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Tests/PackageLoadingTests/PackageBuilderTests.swift [hunk 2]
```diff
@@ -1576,6 +1576,40 @@ class PackageBuilderTests: XCTestCase {
...
+                        "Foo2",
+                    ]),
+                TargetDescription(name: "Foo2"),
+            ]
+        )
+
+        PackageBuilderTester(manifest1, path: AbsolutePath("/Foo"), in: fs) { result in
+            result.checkModule("Foo")
+            result.checkModule("Foo2")
+            result.checkDiagnostic("duplicate target dependency 'Bar' in: Foo")
```
Comment (aciidb0mb3r) [inline Tests/PackageLoadingTests/PackageBuilderTests.swift]: we should make this a little clear:

```
invalid duplicate target dependency declaration 'Bar' in target 'Foo'
```

File: Sources/PackageLoading/PackageBuilder.swift [hunk 3]
```diff
@@ -554,6 +554,13 @@ public final class PackageBuilder {
             // Add the created target to the map or print no sources warning.
             if let createdTarget = target {
                 targets[createdTarget.name] = createdTarget
+
+                // Check for duplicate target dependencies by name
+                let combinedDependencyNames = (createdTarget.dependencies.map { $0.name } + createdTarget.productDependencies.map { $0.0 })
```
Comment (aciidb0mb3r) [inline Sources/PackageLoading/PackageBuilder.swift]: We should move this check into createTarget method.

Comment (michaelmcguire) [reply Sources/PackageLoading/PackageBuilder.swift]: Moved into `createTarget`

File: Tests/PackageLoadingTests/PackageBuilderTests.swift (hunk 2 above)
Comment (michaelmcguire) [reply Tests/PackageLoadingTests/PackageBuilderTests.swift]: Improved error message

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by michaelmcguire** [comment]
@aciidb0mb3r we good to go?

**Comment by neonichu** [comment]
Thanks!

**Pull request merged**